### PR TITLE
WW-4912: scan package-info and class at the same time

### PR DESCRIPTION
**Pull request by lukaszlenart**
This change allows to define `AllowedMethods` in `package-info.java` and all class level and all those annations will be examined to build set of allowed methods.

Fixes [WW-4912](https://issues.apache.org/jira/browse/WW-4912)

### CODE REVIEW

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java [hunk 1]
```diff
@@ -151,4 +152,33 @@ public static String resolvePropertyName(Method method) {
...
     }
+
+    /**
+     * Returns the annotation on the given class or the package of the class. This searches up the
+     * class hierarchy and the package hierarchy for the closest match.
+     *
+     * @param <T>             class type
+     * @param clazz           The class to search for the annotation.
+     * @param annotationClass The Class of the annotation.
+     * @return The annotation or null.
```
Comment (HedjuHor) [inline core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: never return 'null'

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java [hunk 2]
```diff
@@ -151,4 +152,33 @@ public static String resolvePropertyName(Method method) {
...
+            anns.add(ann);
+        }
+
+        ann = clazz.getPackage().getAnnotation(annotationClass);
+        if (ann != null) {
+            anns.add(ann);
+        }
+
+        if (clazz.getSuperclass() != Object.class) {
+            anns.addAll(findAnnotations(clazz.getSuperclass(), annotationClass));
```
Comment (HedjuHor) [inline core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: if someone calls this method with Object.class, the recursive call ends with NPE

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java (hunk 1 above)
Comment (aleksandr-m) [reply core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: never returns `null`

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java [hunk 3]
```diff
@@ -139,19 +140,45 @@ public static String resolvePropertyName(Method method) {
      */
     public static <T extends Annotation> T findAnnotation(Class<?> clazz, Class<T> annotationClass) {
         T ann = clazz.getAnnotation(annotationClass);
-        while (ann == null && clazz != null) {
+        if (ann == null && clazz != Object.class) {
             ann = clazz.getAnnotation(annotationClass);
```
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: As you replaced `while` with `if` then I think you can correct this method as below:

```java
    public static <T extends Annotation> T findAnnotation(Class<?> clazz, Class<T> annotationClass) {
        T ann = null;
        if (clazz != null) {
            ann = clazz.getAnnotation(annotationClass);
            if (ann == null) {
                ann = clazz.getPackage().getAnnotation(annotationClass);
            }
            if (ann == null) {
                ann = findAnnotation(clazz.getSuperclass(), annotationClass);
            }
        }

        return ann;
    }
```

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java [hunk 4]
```diff
@@ -139,19 +140,45 @@ public static String resolvePropertyName(Method method) {
...
+        if (ann != null) {
+            anns.add(ann);
+        }
+
+        ann = clazz.getPackage().getAnnotation(annotationClass);
+        if (ann != null) {
+            anns.add(ann);
+        }
+
+        if (clazz.getSuperclass() != Object.class) {
```
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: What do you feel with this instead:

```java
        ann = findAnnotations(clazz.getSuperclass(), annotationClass);
        if (ann != null) {
            anns.add(ann);
        }
```

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java (hunk 1 above)
Comment (lukaszlenart) [reply core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: Sorry, I forgot to update the JavaDoc :)

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java (hunk 2 above)
Comment (lukaszlenart) [reply core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: I think that's good, calling this function in `Object` is a bad idea anyway.

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java (hunk 4 above)
Comment (lukaszlenart) [reply core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: `findAnnotations` does not return `null` also I don't want to scan above `Object` as it doesn't make sense

**Comment by yasserzamani** [comment]
Thanks! LGTM 👍

**Comment by lukaszlenart** [comment]
@yasserzamani could you hit the button? :)

**Comment by yasserzamani** [comment]
Yeah I was going to push :) but said myself, now I know it works but let see once time again if any further improvements are possible. Now I'm preparing some nice improvements as another PR on this PR 😉 improvements like supporting annotations on interfaces in other packages also :) plus more strict tests e.g. one of my tests showed `findAnotations` will have duplicate results when searches in same package.

**Comment by lukaszlenart** [comment]
Hm... not sure what's the problem with duplications? If we scan the whole class hierarchy having duplications is rather expected.

**Comment by yasserzamani** [comment]
I meant if class A extends B and implements I and A and B and I are in same package P and P is annotated by annotation ANN then it will have three references of ANN in returned list. I felt it's somehow not nice but in another side, now, I also am worry about performance 😟

**Comment by lukaszlenart** [comment]
That's why I would leave it as is and add a task to improve that in 2.6 - tbh, I would like to start a release process of 2.5.15 ASAP

**Comment by yasserzamani** [comment]
👍 Just also please see [here](https://github.com/lukaszlenart/struts/pull/1/files). There are a few but nice improvements which don't need a lot time to review. Please feel free to merge them or just inform me and I'll merge this PR without them if you prefer then we can postpone them and performance improvements to `2.6` :)

**Comment by yasserzamani** [comment]
Great work 👍 thanks!

**Pull request merged**